### PR TITLE
Update kubeletConfigFile toggle and fix

### PR DESCRIPTION
**Pull request by bowang-666**
Rename DynamicKubelet to KubeletConfigFile since it's really not dynamic.
Fix issues configuring TopologyManager and FailSwapOn.
Disable DynamicKubeletConfig feature gate if using kubelet config file.

### CODE REVIEW

File: pkg/agent/baker.go [hunk 1]
```diff
@@ -264,8 +264,8 @@ func getContainerServiceFuncMap(config *NodeBootstrappingConfiguration) template
 			return profile.CustomLinuxOSConfig.TransparentHugePageDefrag
 		},
 		"ShouldConfigSwapFile": func() bool {
-			// only configure swap file when FailSwapOn is true and SwapFileSizeMB is valid
-			return profile.CustomKubeletConfig != nil && profile.CustomKubeletConfig.FailSwapOn != nil && *profile.CustomKubeletConfig.FailSwapOn &&
+			// only configure swap file when FailSwapOn is false and SwapFileSizeMB is valid
```
Comment (xuto2) [inline pkg/agent/baker.go:267]: what is the default behavior of this FailSwapOn ? true or false?

File: pkg/agent/utils.go [hunk 2]
```diff
@@ -371,21 +371,23 @@ func GetOrderedKubeletConfigFlagString(k *datamodel.KubernetesConfig, cs *datamo
...
 
-// IsDynamicKubeletEnabled get if dynamic kubelet is supported in AKS and toggle is on
-func IsDynamicKubeletEnabled(cs *datamodel.ContainerService, dynamicKubeletToggleEnabled bool) bool {
+// IsKubeletConfigFileEnabled get if dynamic kubelet is supported in AKS and toggle is on
+func IsKubeletConfigFileEnabled(cs *datamodel.ContainerService, profile *datamodel.AgentPoolProfile, kubeletConfigFileToggleEnabled bool) bool {
 	// TODO(bowa) remove toggle when backfill
-	return dynamicKubeletToggleEnabled && cs.Properties.OrchestratorProfile.IsKubernetes() && IsKubernetesVersionGe(cs.Properties.OrchestratorProfile.OrchestratorVersion, "1.14.0")
+	// If customKubeletConfig or customLinuxOSConfig is used (API20201101 and later), use kubelet config file
+	return profile.CustomKubeletConfig != nil || profile.CustomLinuxOSConfig != nil ||
+		kubeletConfigFileToggleEnabled && cs.Properties.OrchestratorProfile.IsKubernetes() && IsKubernetesVersionGe(cs.Properties.OrchestratorProfile.OrchestratorVersion, "1.14.0")
```
Comment (xuto2) [inline pkg/agent/utils.go:379]: nit: can we add a round bracket to make it clearer ?
 (kubeletConfigFileToggleEnabled && cs.Properties.OrchestratorProfile.IsKubernetes() && IsKubernetesVersionGe(cs.Properties.OrchestratorProfile.OrchestratorVersion, "1.14.0"))

File: pkg/agent/baker.go (hunk 1 above)
Comment (bowang-666) [reply pkg/agent/baker.go:267]: Default is true. It makes kubelet fail to start if node has swap. So we have to explicitly turn it off to use swap.
https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/#:~:text=--fail-swap-on

File: parts/linux/cloud-init/artifacts/kubelet.service [hunk 3]
```diff
@@ -26,7 +26,7 @@ ExecStart=/usr/local/bin/kubelet \
         --node-labels="${KUBELET_NODE_LABELS}" \
         --v=2 {{if NeedsContainerd}}--container-runtime=remote --runtime-request-timeout=15m --container-runtime-endpoint=unix:///run/containerd/containerd.sock{{end}} \
         --volume-plugin-dir=/etc/kubernetes/volumeplugins \
-        {{- if IsDynamicKubeletEnabled}}
+        {{- if IsKubeletConfigFileEnabled}}
         --config /etc/default/kubeletconfig.json \
         {{- end}}
         $KUBELET_FLAGS \
```
Comment (xuto2) [inline parts/linux/cloud-init/artifacts/kubelet.service:32]: Just curious , can we combine --config /etc/default/kubeletconfig.jso into KUBELET_FLAGS? I think we don't have a clear principal as where the kubelet flags should go

File: pkg/agent/baker.go (hunk 1 above)
Comment (xuto2) [reply pkg/agent/baker.go:267]: ok. so we should have validation in RP to error out if failswapon is true and swapFileSizeMB is not nil

File: pkg/agent/utils.go (hunk 2 above)
Comment (bowang-666) [reply pkg/agent/utils.go]: fixed

File: parts/linux/cloud-init/artifacts/kubelet.service (hunk 3 above)
Comment (bowang-666) [reply parts/linux/cloud-init/artifacts/kubelet.service:32]: Technically we can, but I just want to call it out specifically here.

File: pkg/agent/baker.go (hunk 1 above)
Comment (bowang-666) [reply pkg/agent/baker.go:267]: Yes.. there's actually quite a lot of validation we need to do

File: parts/linux/cloud-init/artifacts/kubelet.service (hunk 3 above)
Comment (xuto2) [reply parts/linux/cloud-init/artifacts/kubelet.service:32]: got you.

**Comment by xuto2** [comment]
also remember to revert the duplicated merging logic RP code when this change is vendored to RP

File: pkg/agent/types.go [hunk 4]
```diff
@@ -42,7 +42,7 @@ type NodeBootstrappingConfiguration struct {
 	UserAssignedIdentityClientID  string
 	ConfigGPUDriverIfNeeded       bool
 	EnableGPUDevicePluginIfNeeded bool
-	EnableDynamicKubelet          bool
```
Comment (xuto2) [inline pkg/agent/types.go:45]: FYI https://github.com/Azure/AgentBaker/pull/410/files# this PR removes pkg/agent/types.go

**Comment by bowang-666** [comment]
> also remember to revert the duplicated merging logic RP code when this change is vendored to RP

sure

**Pull request merged**